### ISSUE TITLE
ROCm: FP32 GEMM throughput capped at 3.1 TFLOPS (21% of theoretical 14.85)

### ISSUE THREAD

**Issue by bong-water-water-bong** [PrismML-Eng/Bonsai-demo#91]
## Environment
- GPU: Radeon 8060S (gfx1151, Strix Halo)
- ROCm 7.2.4
- PyTorch 2.13.0+rocm7.2

## Issue
FP32 GEMM throughput maxes out at ~3.1 TFLOPS regardless of matrix size (512x512 through 4096x4096), which is only ~21% of the theoretical peak of 14.85 TFLOPS FP32.

In contrast, FP16 GEMM achieves 25.9 TFLOPS sustained, which is 87% of the FP16 theoretical peak (29.7 TFLOPS).

```
FP32 4096x4096:  3.09 TFLOPS  (21% of 14.85 peak)
FP16 4096x4096: 33.09 TFLOPS  (111% of 29.7 peak — may include dual-issue)
```

## Impact
Any FP32-heavy workload is severely bottlenecked. Most inference frameworks default to FP32 for certain operations (e.g., attention, softmax, layer norms). This creates an unexpected performance cliff.

## Suspected Cause
RDNA 3.5 (gfx1151) may have half-rate FP32 in the shader array, or the ROCm tensor/GEMM path isn't selecting the optimal kernel for FP32 on this architecture. FP16 gets full rate via packed math.

## Workaround
Always use FP16 for both weights and activations. Set model dtype to `torch.float16`.

**Comment by bong-water-water-bong** [comment]
## DIAGNOSED — Architectural limit

FP32 throughput is capped at ~3.1 TFLOPS (21% of 14.85 theoretical peak). This is consistent with **RDNA 3.5's half-rate FP32** on the shader array — FP32 operations run at half the rate of FP16 on the same execution units.

**Comparison:**
```
Test                     FP16        FP32       Ratio
GEMM 1024x1024       38.0 TFLOPS   2.9 TFLOPS   13x
GEMM 4096x4096       33.1 TFLOPS   3.1 TFLOPS   10.7x
```

FP16 achieves its full rate via packed math (2× FP16 per FP32 operation slot). FP32 can only use one lane per slot, effectively halving throughput. The 3.1 TFLOPS observed is consistent with ~20 CUs (of 40) being exposed to PyTorch at FP32 with ~2.9 GHz clock with some scheduling overhead.

**Not a bug** — this is expected behavior for RDNA 3.5 (`gfx1151`). The solution is:

1. **Use FP16 exclusively** for all workloads (weights AND activations)
2. Use quantization (Q4_K_M, IQ1_S, etc.) for even faster inference
3. Mixed precision training with FP16 master weights

**Already the default behavior** in PyTorch (`torch.float16` / `torch.bfloat16`) and in llama.cpp with quantized models.
